### ISSUE TITLE
`V6`: 🎉 Group Rewrite

### ISSUE THREAD

**Issue by ShaMan123** [fabricjs/fabric.js#7670]
## V6
Hi everyone,
**fabric is moving forward!**
We have 2 main goals for v6:
1. js/ts migration which is covered here #7596 
2. Group rewrite 🎉

You can see #8316

## Group Rewrite
**This is a tracking issue for the group rewrite**, read below for more detail including the current state of things and gist.
We aim to fix and support long desired features including:
1. Nested selection - **DONE**
2. Resizing and customized layout - **DONE**
4. Transform and coordinate system (which is relative to the parent group) - **DONE**

### Motivating Issues/Discussions
#7473 https://github.com/fabricjs/fabric.js/pull/7316 https://github.com/fabricjs/fabric.js/issues/6776 https://github.com/fabricjs/fabric.js/discussions/7136 https://github.com/fabricjs/fabric.js/discussions/7299 https://github.com/fabricjs/fabric.js/issues/7130 https://github.com/fabricjs/fabric.js/issues/7142 #7449 and anything with a `group` tag

### How do you fit in?
1. You can become an alpha tester (we aim to publish an alpha tag to npm regularly). 
**IMPORTANT NOTICE**: all changes made to the v6 branch are experimental and can be removed or changed without notice and without proper support (including supported features from v5). fabric might change drastically between commits. Meaning if you want in - it's on you.

You can join the effort by [**contributing**](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)

**This description is updated continuously with relevant information**
You don't need to scroll and read endless comments (unless you are a maintainer :))

---

## Current State
Fixed logic across fabric for nested objects to be selectable.
Group has been completely rewritten and Layer has been introduced. As part of the rewrite group is now a layout engine, you can use it to customize layout of objects.
In general the rewrite allows any object to be selectable and interactive if it is referenced in the `_objects` array of it's parent and it references it's parent via the `group` property. If you build custom objects with nesting this is what you are looking for.

### Usage
It is advised to use `set` on group/objects from now on (until #7596 is done, then it might change due to setters)
Changing `subTargetCheck` or `interactive` enables/disables crucial functionality so for these props it is not an advice but a **MUST**.
```js
new fabric.Group(objects, {
  subTargetCheck: true,
  interactive: true   //  enables selecting subtargets
});
```

### Progress 
- [x] #7858 
- [x] #7859 
- [x] #7916 
- [x] #8000 
- [ ] #7860 *needs porting to v6*
- [x] #7861 
- [x] #8004 
- [ ] #8034 *needs porting to v6*
- [x] #8461
- [x] ~~#7878~~
- [x] ~~#7882~~
- [x] #8665 
- [ ] #7893 *needs porting to v6*
- [ ] #7901 *needs porting to v6*
- [x] #8951 

### Experimental

- [ ] #7879 
- [ ] #7880 

#### Triggering Layout

#9148 extracts layout management to a standalone class

`triggerLayout` forces a group to recalculate its bounding box and reposition its objects.
You can define a layout strategy and assign it to the layout manager.
Read the code, it has comments explaining stuff and types that make it clear what you need to return from the method.
Look at the [modified trigger](https://github.com/fabricjs/fabric.js/blob/a39e01cf5ceae66809272051f77319345d8492d2/src/shapes/Group.ts#L347-L348), it is possible adding more triggers.
However I will promote an overall solution.
[#7882](https://github.com/fabricjs/fabric.js/pull/7882/files#diff-216444c1bc3837d560c8241c4d9643b6de8c147095081c5b38715b9e058c24cbR454-R471) is a POC, follow [the progress trigger](https://github.com/fabricjs/fabric.js/blob/91a7f533380f46e6684ff399f68e92025bcf6d19/src/shapes/group.class.js#L464-L470) 

### BREAKING
- all awkward methods such as `addWithUpdate` are removed including `toGroup`/`toActiveSelection`
- renamed object stacking methods on canvas (e.g. `moveTo` -> `moveObjectTo`)
- strict object tree - **EXTREMELY BREAKING** because it steals under the radar
   - an object can have only one parent
   - no need to call `remove` before adding to another group, fabric handles it internally
   - no need (and discouraged) to use awkward active selection code for removing an object
   - canvas methods still need to enforce a strict check 
- if you hacked group prior to the rewrite you will probably need to rewrite your logic

## Pitfalls

All object origin methods (e.g. `translateToOriginPoint`) are relative methods.
You should use them **ONLY** in the correct plane and with points that belong to the relative plane.
IMO it makes them worthless.
#8767 changes that.

## BUGS

If you encounter bugs related to group or if you have a feature request please submit a relevant ticket.
**Notice that bug reports of versions below v6 will be completely disregarded and closed.** This was hard work, put effort into your repro so it is in v6.

## Working Examples
- [**JSFiddle**](http://jsfiddle.net/qxdk4h9L/2/)
- [**Layer**](http://jsfiddle.net/14rqvut0/)

~~**Using `v6!` stale branch**~~
~~- [**CodeSandbox**](https://codesandbox.io/s/jumshw) or [**JSFiddle**](http://jsfiddle.net/Lxjegow3/8/)~~
~~- [**Layer**](https://codesandbox.io/s/fabric-app-forked-cxcz3) or in [**JSFiddle**](http://jsfiddle.net/6xjszqtg/1/)~~

## To Do

- [x] refactor selection logic
   - [x]  use `dblClick` to select nested objects? create mechanism for dev to customize selection? for now override `searchPossibleTargets`
   - [x] refactor to enable n levels of nesting (logic is adequate for 2)  https://github.com/fabricjs/fabric.js/blob/109efe52d1ae6de780389c0dbfefbaf051646a21/src/canvas.class.js#L422 
   - [x] look into active selection - creating from nested objects (deprecate/refactor `_chooseObjectsToRender`)
   - [ ] z-index of relatives when deselecting an object and selecting another - https://codesandbox.io/s/yiliv
   - [ ] alt-selection for selecting objects under top object with another click - [**POC**](https://github.com/fabricjs/fabric.js/compare/v6-nested-selection...v6-nested-selection-alt)
- [ ] @asturur  rethink `preserveObjectStacking` in favor of a drag image/ghost - can wait/stale due to #7878 
- [x] text selection in scaled and rotated group is wrong (`getLocalPointer` 8bf7f92)
- [x] bounding box and controls of object under transformed group (#7758 )
- [x] ~~phase out coords (can be delayed, now that bbox and controls are fixed)~~ #8767 
- [x] update tests
- [ ] refactor object caching - #7874 
   - [ ] object caching in some cases leaves traces (probably caused by wrong size when clearing cache)
   - [ ] caching edge case - nested absolute clip path (see [fiddle](http://jsfiddle.net/prw6g1x5/), try moving group, image should be clipped). Possible fix is to handle a registry of children with abs clip paths **OR** modify the `shouldCache` method
   - [ ] fix group's caching behavior + shadow logic - #7454 
   - [ ] rethink `_renderObjects` - nested active object will fail with `perPixelTargetFind` because it will not get rendered by group (add logic to `canvas.isTargetTransparent`) #7863 
   - [ ] rethink `ActiveSelection` caching - why not?
- [x] support layers under groups #7860 
- [x] ActiveSelection.toGroup use `removeAll` instead
- [x] `getObjectsBoundingBox` - needs work to exactly fit when group is rotated
- [x] initial layout inconsistent with old group
- [x] initial layout - supporting `originX/Y` 407023cc and `angle` #7867 
- [x] initial layout ( compare against the working examples of v6!) (Layer/Group)
- [ ] Layer - respect originX/originY?
- [x] Group `centerObject` methods:
    -  `object.setRelativeCenterPoint(new Point())`
    -  `object.setRelativeCenterPoint(new Point(object.getRelativeCenterPoint().x, 0))`
- [ ] should `clipPath` become evented/interactive? #7879 
- [x] `groupSVGElements`
- [x] ~~support fill, stroke~~ ~~`backgroundColor` - might be replaced with `fill-parent` layout for object~~ will not be supported by fabric, see https://github.com/fabricjs/fabric.js/pull/7859#discussion_r846930838
- [ ] remove object selection monitors in favor of a simple check `canvas.getActiveObjects().includes(obj)`
- [x] canvas/group `removed` event dereferencing inconsistency https://github.com/fabricjs/fabric.js/issues/7859#issuecomment-1109377515_ - fixed by #8000
- [x] ~~`flipX/Y` on group and/or nested objects is a complete mess~~ #8767 

## Look
Video is running in original speed
![tiger](https://user-images.githubusercontent.com/34343793/153901183-e6de11cf-c0fb-42a9-aac3-bea6eebabf24.gif)

## Cutting Edge
The blue circles are rendered by a textbox and because of the rewrite with a few lines of code I got this working
Pay no attention to caching of course

This is **EXTREMELY** exciting because it means that anyone could create custom objects subclassing whatever object they need and have nested selection work out of the box!

![ezgif com-gif-maker (13)](https://user-images.githubusercontent.com/34343793/158232754-3ba52ba0-8d9c-45ba-9801-3364def61821.gif)

**Comment by ShaMan123** [comment]
First milestone #7858 🤟

**Comment by ShaMan123** [comment]
First milestone is merged!! 🥳🥳

**Comment by ShaMan123** [comment]
updated `v6-alpha` branch, now contains all group PRs + 
- #7802 
- #7907 
- #7713

**Comment by ShaMan123** [comment]
Second milestone merged!

**Comment by ShaMan123** [comment]
4th milestone merged!

**Comment by ShaMan123** [comment]
@asturur I think we can drop coods in favor of `getCenterPoint` + `_getTransformedDomensions`
What do you think?

**Comment by ShaMan123** [comment]
Group is now in beta!

**Comment by dubem-design** [comment]
How are things going?

**Comment by ShaMan123** [comment]
We are prioritizing TS so the rest of the open prs will find their way in after that.
You can work on the alpha branch if you need to

**Comment by ShaMan123** [comment]
I am pretty sure some of the prs will be rewritten to accomodate better caching logic

**Comment by Sikhi** [comment]
@ShaMan123 really a lot of branches, to hard to follow latest changes. Good description in first post but where latest code is not clear

**Comment by ShaMan123** [comment]
Under the **Alpha testing** section it is very clear IMO

**Comment by Sikhi** [comment]
> Under the **Alpha testing** section it is very clear IMO

yes, but at 16 Jun you said that `Group is now in beta!`

**Comment by ShaMan123** [comment]
It has nothing to do with the naming of the branch

**Comment by dubem-design** [comment]
Just checking up ✌️

**Comment by ShaMan123** [comment]
#8316

**Comment by ShaMan123** [comment]
Update for Groupies
Group has been migrated to TS along with ActiveSelection #8455.
This means I can port open PRs to v6 so that they can enter the merging queue (which isn't short).
We plan to delegate the layout mechanism to a standalone class as part of the migration and for ease of customization and subclassing. I am considering adding a small feature that anchors objects to a given origin during layout. This should make logic much more simple and accessible (It takes me a lot of time to write more layout logic and I might be the only one that does it currently which says bad things about design). 
I am now convinced that we should introduce a mechanism that will support interactive group with the handling of caching instead of the ugly workarounds I did to make it work (I now consider them as mere POCs) so expect it to change, especially that #8298 is a valid POC and if merged will have a real impact on fabric.

**Comment by enriqueerrando** [comment]
Hi,

I was wondering if V6 will change how toDatalessJSON structures data.

The thing is I'm building a designer tool based on FabricJS and the user will be able to save to the db the canvas data via toDatalessJSON to retreive it later with loadFromJSON. Will V6 be able to load that data after being exported with V5?

Thanks in advance!

**Comment by ShaMan123** [comment]
@enriqueerrando `toDatalessJSON | toDatalessObject` are methods that are used after parsing from svg. It has been changed in group. It used to have this strange option that the objects property was a link to an svg file. That was removed. 
The only object that actually implements this method is Path (IMO we should remove that as well and deprecate the entire concept).
If you aren't using svg parsing and Path in particular, using `toObject` will have the same result.

**Comment by timoostrich** [comment]
> @enriqueerrando `toDatalessJSON | toDatalessObject` are methods that are used after parsing from svg. It has been changed in group. It used to have this strange option that the objects property was a link to an svg file. That was removed. The only object that actually implements this method is Path (IMO we should remove that as well and deprecate the entire concept). If you aren't using svg parsing and Path in particular, using `toObject` will have the same result.
> 
> So answering your question, you might need to re-export objects from v5 using `toObject`. I think it is a good idea we expose a link that does that... you load a json, it downloads a new one. Sure, it will be done.

Glad to hear that as I also created a design tool where users saved thousand of files till now... :D

**Comment by ShaMan123** [comment]
I don't store fabric raw output in db. I act on it, too much junk data.

**Comment by timoostrich** [comment]
Me neither, I'm saving it as a gzipped file on the server.
What do you do with the data?

**Comment by ShaMan123** [comment]
Depends on business logic. I am not for gzipped files since then you have to fetch all of it instead of bits of interest. But that might be suitable for your app.

**Comment by ShaMan123** [comment]
I gave it another thought.
Since we are planning to release a registry concept in v6 (POC #8282) I think (depending on the final design) it would be possible to add a custom/legacy resolver that patches differences in serialization between versions without any elaborate workarounds.

**Comment by enriqueerrando** [comment]
Hi @ShaMan123 and @timoostrich, thanks for answering ;)

No, I do not save data to the db neither, I do save it to a file too (not gzipped though), but for the example obove I thought was simpler to say so because thought that was what the majority was doing, but I can see now it isn't ;)

I use loadSVGFromURL then apply groupSVGElements to them.

If I use toObject woudn't that save all SVG code insted of referencing the URL? I was using toDatalessJSON to reduce the size of the exported data.

> I gave it another thought.
> Since we are planning to release a registry concept in v6 (POC https://github.com/fabricjs/fabric.js/pull/8282) I think (depending on the final design) it would be possible to add a custom/legacy resolver that patches differences in serialization between versions without any elaborate workarounds.

That would be great, thanks! But since the V6 release is so close, I think I'll just wait for it to make sure all of my code works fine with it.

**Comment by ShaMan123** [comment]
Fabric is almost fully migrated, Group is doing **GREAT**! #8461 just merged.

Remaining work:
- strict tree safeguards
- selection logic
- features

**Comment by enriqueerrando** [comment]
Those are great news!! Thanks for the hard work!

**Comment by timoostrich** [comment]
Really looking forward to it, thank you so much for your effort!

**Comment by dubem-design** [comment]
Wow Great work. I can't wait

**Comment by enriqueerrando** [comment]
Hi @ShaMan123,

How is V6 going? I've seen there are some tasks left to finish it, but don't know how that translates in time.

Thanks a lot for your work!

**Comment by ShaMan123** [comment]
Yesterday or something @asturur managed to stabalize the build.
So unofficially we are in beta. A few days, a week, until we roll out the official version.

**Comment by enriqueerrando** [comment]
Woooow! I didn't expect that answer, hahaha!! Great news then!! Thanks for everything u're doing ;)

**Comment by enriqueerrando** [comment]
Hi @ShaMan123 !

How is the beta testing going?

Thanks in advance!

**Comment by ShaMan123** [comment]
#8622

**Comment by enriqueerrando** [comment]
It's looking good ;)

**Comment by ShaMan123** [comment]
#8665 has been merged meaning that nested mutliselection is now supported. 
Still a few issues with group caching that break mutliselection.
Workaround by disabling group caching in such a case.

**Comment by enriqueerrando** [comment]
Great! Thanks for keepping us updated!

I'm just waiting for the V.6 release to try it all together, because the web app I'm building needs several things that may change from 5 to 6, so I cannot continue 'till the update ;)

Keep up the good work!! U're almost done!

**Comment by abdussami-tayyab** [comment]
Is there something in v6 that can help with this usecase:
- When a group is double-clicked (I will use `mouse:dblclick` event of Canvas), group must be temporarily ungrouped, manipulation on shapes can be done by user, then after that it must be re-grouped once the user presses elsewhere on canvas.

cc: @ShaMan123

**Comment by ShaMan123** [comment]
Group was redesigned to be simple (apart from one mistake I hope to fix soon)
So you simply call group.removeAll and then to regroup you can call group.add toadd the objects back to the original group or create a new one.
That is it.

**Comment by abdussami-tayyab** [comment]
There is an issue with the way `Group` is working for me. The interactivity (that the project I'm working on **massively** will rely on) is great, subtargets also work well. But there's quite a glitchy line when I move the group around. [Here](https://www.loom.com/share/930377cbcce34efba14f4a6a59afaf15) is the video.

And here is the [CodeSandbox](https://codesandbox.io/s/cocky-drake-few1sf).

This is more or less the same way I did it with v5 (which is working), but since v6 fits exactly my use-case I want to go with it. But this is causing quite an issue, especially when the background is white (which we'll go with eventually).
Another thing is that the line when moved around via it's control circle, part of it vanishes which is outside the bounding box's bounds. 

<img width="325" alt="Screenshot 2023-05-18 at 11 28 17 PM" src="https://github.com/fabricjs/fabric.js/assets/7189666/e854e283-701c-47f1-bae7-9d7e4c05c2d5">

**Comment by abdussami-tayyab** [comment]
Just wanted to update about the above comment in case it helps: I am not facing the above issue when I set the `hasBorders` of the `Line` object to `false`.

**Comment by ShaMan123** [comment]
This is a caching issue.
Perhaps it is not well documented. Caching is still a working progress.
Disable group caching. Or mark it is dirty...

**Comment by ShaMan123** [comment]
Rect.selectable = false?

**Comment by abdussami-tayyab** [comment]
It will make the rect unselectable/non-interactive which is also something I don't want. I just tried this use-case in Canva, and in Canva if I create a rectangle of 50x50 and place a circle of 30x30 inside it, then group both of them. Then when I select the group to move, the whole group moves and I am not able to move the rectangle or circle although I can change their attributes.

I wonder what event listeners we can use, for example if a user moves a group without Shift key pressed, we deal with the group but if Shift key is pressed, we can select the individual object inside the group. If this is an advanced topic for us, we can have a look at it later, but important to at least think about it because we are sure this is not an unusual case.

**Comment by ShaMan123** [comment]
Did you try toggling the selectable prop in correlation to the shift key?
You can also try to fiddle with onSelect, onDeselect and the related events

**Comment by abdussami-tayyab** [comment]
Sorry but can you please point me to that in the code from where I see how
it works? Thanks.

**Comment by abdussami-tayyab** [comment]
@ShaMan123 I tried setting the `selectable` attribute of objects inside a group, the problem with that now is that moving the group now does nothing because the underlying object which has `selectable: false`.

Is there any way we can set `interactive` only on `IText` objects and not on any other shape?

**Comment by abdussami-tayyab** [comment]
@asturur can you help with my comment above? If I can make some objects in a group interactive and some not, that would be golden for the project I'm working on. If this is not in the plan for now, that would work as well if I can get to know. Thanks for this great package!

cc: @ShaMan123

**Comment by ShaMan123** [comment]
Open a dedicated discussion/issue
Provide a reproduction

**Comment by asturur** [comment]
@abdussami-tayyab link the open issue when you open it.
The interactive group feature wasn't planned with all those customization in mind, so some of them needs to be evaluated, thought, and built from scratch.

**Comment by abdussami-tayyab** [comment]
@asturur do we have a boilerplate JSFiddle or CodeSandbox that I can use to replicate my issue?
Sorry about commenting here, I understand it breaks the purpose of this thread. I'll delete once you tell me how I can add code to JSFiddle or CodeSandbox (or any other tool).

Thanks!

**Comment by StringKe** [comment]
Any update on the implementation of this [DEMO](http://fabricjs.com/manage-selection) in V6?

**Comment by StringKe** [comment]
> Are you referring to toGroup/toActiveSelection?
> 
> `toActiveSelection`
> 
> ```ts
> canvas.getActiveSelection.add(...group.removeAll())
> //or
> canvas.setActiveObject(new ActiveSelection(group.removeAll())
> // remove group
> ```
> 
> `toGroup`
> 
> ```ts
> new Group(canvas.getActiveObjects())
> ```

Thanks, your code is very clean, I used a lot of code to achieve a similar result.

**Comment by ShaMan123** [comment]
> Thanks, your code is very clean, I used a lot of code to achieve a similar result.

Yes, v6 is a major step forward for the repo and the group rewrite especially

**Comment by StringKe** [comment]
`"fabric": "^6.0.0-beta15"`  A lot of typescript types are incorrect in this version, in my local project there are up to 200 type checking errors in fabric! 

<details><summary>Codesanbox</summary>
<p>

```
✔ Browser application bundle generation complete.

Error: node_modules/fabric/dist/src/filters/HueRotation.d.ts:14:22 - error TS2417: Class static side 'typeof HueRotation' incorrectly extends base class static side 'typeof ColorMatrix'.
  The types of 'defaults.mainParameter' are incompatible between these types.
    Type 'keyof HueRotation | undefined' is not assignable to type 'keyof ColorMatrix | undefined'.

14 export declare class HueRotation extends ColorMatrix {
                        ~~~~~~~~~~~


Error: node_modules/fabric/dist/src/shapes/Path.d.ts:39:17 - error TS2300: Duplicate identifier 'path'.

39     constructor(path: TComplexPathData | string, { path, left, top, ...options }?: Partial<Props>);
                   ~~~~


Error: node_modules/fabric/dist/src/shapes/Path.d.ts:39:52 - error TS2300: Duplicate identifier 'path'.

39     constructor(path: TComplexPathData | string, { path, left, top, ...options }?: Partial<Props>);
                                                      ~~~~
```

</p>
</details> 

Codesanbox https://codesandbox.io/p/devbox/charming-williams-4qrkfz

<details><summary>Local angular 16.2.10</summary>
<p>

```
Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/Pattern/Pattern.ts:177:34 - error TS2339: Property 'width' does not exist on type 'CanvasImageSource'.
  Property 'width' does not exist on type 'VideoFrame'.

177           : ifNaN((patternSource.width as number) / width, 0),
                                     ~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/Pattern/Pattern.ts:181:34 - error TS2339: Property 'height' does not exist on type 'CanvasImageSource'.
  Property 'height' does not exist on type 'VideoFrame'.

181           : ifNaN((patternSource.height as number) / height, 0);
                                     ~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/Pattern/Pattern.ts:185:50 - error TS2339: Property 'width' does not exist on type 'CanvasImageSource'.
  Property 'width' does not exist on type 'VideoFrame'.

185       `<image x="0" y="0" width="${patternSource.width}" height="${
                                                     ~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/Pattern/Pattern.ts:186:23 - error TS2339: Property 'height' does not exist on type 'CanvasImageSource'.
  Property 'height' does not exist on type 'VideoFrame'.

186         patternSource.height
                          ~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/Canvas.ts:112:11 - error TS2564: Property '_isClick' has no initializer and is not definitely assigned in the constructor.

112   private _isClick: boolean;
              ~~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/Canvas.ts:1599:3 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'SelectableCanvas<CanvasEvents>'.

1599   clear() {
       ~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/Canvas.ts:1607:3 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'SelectableCanvas<CanvasEvents>'.

1607   destroy() {
       ~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/DOMManagers/CanvasDOMManager.ts:92:3 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'StaticCanvasDOMManager'.

92   setDimensions(size: TSize, retinaScaling: number) {
     ~~~~~~~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/DOMManagers/CanvasDOMManager.ts:98:3 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'StaticCanvasDOMManager'.

98   setCSSDimensions(size: Partial<CSSDimensions>): void {
     ~~~~~~~~~~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/DOMManagers/CanvasDOMManager.ts:104:3 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'StaticCanvasDOMManager'.

104   cleanupDOM(size: TSize) {
      ~~~~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/DOMManagers/CanvasDOMManager.ts:116:3 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'StaticCanvasDOMManager'.

116   dispose() {
      ~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/SelectableCanvas.ts:274:10 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'StaticCanvas<EventSpec>'.

274   static ownDefaults: Record<string, any> = canvasDefaults;
             ~~~~~~~~~~~


Error: node_modules/.pnpm/fabric@6.0.0-beta15/node_modules/fabric/src/canvas/SelectableCanvas.ts:276:10 - error TS4114: This member must have an 'override' modifier because it overrides a member in the base class 'StaticCanvas<EventSpec>'.
```

</p>
</details>

**Comment by abdussami-tayyab** [comment]
Can someone please add how to setup using ReactJS? It would be very
important and won’t take much time.

**Comment by StringKe** [comment]
> Can someone please add how to setup using ReactJS? It would be very important and won’t take much time.

How much does this have to do with react? 

angular tries to detect types in node_module.

If the project uses vite's [react + typescript](https://codesandbox.io/p/devbox/cool-cannon-2ywv6s) templates these issues do not occur.

If a project uses nx.dev for react and is packaged with vite, it will have type issues. nx seems to continue to do some sort of type checking after the vite compilation. nx has a configuration `skipTypeCheck`

https://nx.dev/nx-api/vite/executors/build#skiptypecheck

**Comment by ShaMan123** [comment]
Group rewrite has completed a while back and the LayoutManager has been introduced.

#8034 hasn't been merged so look out when adding objects to a group that belonged to a canvas. Call `Canvas#remove` to be sure until that is fixed. The rest of the add operations should manage removing objects from previous group for you.

The parent ref has been introduced to complement the existing group ref.
Basically the group ref is what defines the plane the object is in while the parent ref describe where the object is in the object tree. 
When selected as part of an ActiveSelection, a nested object (one under some group n levels) will have its group ref set to the active selection and its parent ref set to the group it came from. 
When not selected both refs point to the group. 
This is needed for a couple of things including the ability to move the object back to its parent after deselected and rendering it in the case of `preserveObjectStacking`.

There are pending issues with the transform system and the caching system that will cause bugs when using groups such as:
 #8743 This bug is caused because fabric uses decomposed values instead of pure matrix multiplication.
Look out for bugs coming from a combination of shadows and caching/clipPaths.

I suggest anyone needing nested objects and groups to familiarize with `sendPointToPlane`, `sendVectorToPlane` and `sendObjectToPlane`. They do the heavy up lifting of the linear algebra for you. Once you understand how they work it becomes easy to do complex linear algebra in your code.
Be mindful of the plane you are working in, whether it is the viewport, the scene, the parent plane, the object plane, some other internal object plane (fabric defines many of these but none are easy to grasp, such as pathOffset) or weirder ones.
Doing so will save dev hours and make your logic robust for nesting.
